### PR TITLE
Task/sdk 3814/user attribute triggering

### PR DESCRIPTION
**Pull request by Anush-Shand**

### CODE REVIEW

File: clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java [hunk 1]
```diff
@@ -438,7 +440,7 @@ public void run() {
                 }
                 synchronized (PROFILE_FIELDS_IN_THIS_SESSION) {
                     try {
-                        JSONObject profile = dbAdapter.fetchUserProfileById(accountID);
+                        JSONObject profile = dbAdapter.fetchUserProfileByAccountIdAndDeviceID(accountID, deviceInfo.getDeviceID());
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java:443]: is the device id guaranteed to be present when the code flow comes here?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java [hunk 2]
```diff
@@ -54,13 +54,15 @@ public class LocalDataStore {
 
     private final String eventNamespace = "local_events";
 
+    private final DeviceInfo deviceInfo;
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java:57]: we need to check any cases if device info async block is not completed and we fetch deviceId premtively.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/db/DBAdapter.kt [hunk 3]
```diff
@@ -132,26 +132,68 @@ internal class DBAdapter(context: Context, config: CleverTapInstanceConfig) {
...
+            dbHelper.readableDatabase.query(
+                tName,
+                null,
+                "${Column.ID} = ? AND ${Column.DEVICE_ID} = ?",
+                arrayOf(accountId, deviceId),
+                null,
+                null,
+                null
+            )?.use { cursor ->
+                if (cursor.moveToFirst()) {
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/db/DBAdapter.kt:187]: make sure `cursor.moveToFirst()` does not throw exception if no rows are returned.

Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/db/DBAdapter.kt:187]: False is returned if cursor is empty

File: clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java (hunk 2 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java:57]: Yess. Added to testing scope todos

File: clevertap-core/src/main/java/com/clevertap/android/sdk/login/LoginController.java [hunk 4]
```diff
@@ -132,6 +131,8 @@ public Void call() {
                     } else {
                         deviceInfo.forceNewDeviceID();
                     }
+
+                    localDataStore.changeUser();
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/login/LoginController.java:135]: is this changed cause we need deviceid?

**Review by CTLalit — COMMENTED** [review]
Check comments.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/login/LoginController.java (hunk 4 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/login/LoginController.java:135]: Yess

**Comment by Anush-Shand** [comment]
@CTLalit @piyush-kukadiya 

DeviceID null possibilitites

**DeviceId not initialised yet and onUserLogin performed**
- onUserLogin will return before doing any operation because of deviceID null check

**DeviceId not initialised yet and pushProfile/increment etc. is called**

- pushProfile will not be able to cache the guid (same behaviour as existing if deviceId is not initialised yet)
- The event will be queued as is
- New updated profile will be saved in cache but **`not in db as profile is not inflated yet`**


DeviceId initialised but pushProfile / onUserLogin called before profile is inflated in cache
- Since local cache is synchronised, it will wait for inflating and only then move to evaluation and 
computing new values

**New deviceID will always be generated first before inflating profile for asynProfieSwitchUser**

File: clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java (hunk 1 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java:440]: Yess

**Comment by Anush-Shand** [comment]
Reopening with another branch

**Pull request closed without merging**